Add calibrate_decision_threshold

Innixma opened the pull request:
*Issue #, if available:*
resolves #1038, #1515 
pre-req for #3112 

*Description of changes:*
Add `TabularPredictor.calibrate_decision_threshold` which allows to optimize a given metric's decision threshold for predictions to strongly enhance the metric score.

The below code and output showcase how this logic can reliably improve metric scores on a variety of metrics, particularly `f1` and `balanced_accuracy`.

<details>
  <summary>Example Code</summary>

```python
from autogluon.tabular import TabularPredictor, TabularDataset


if __name__ == '__main__':
    path_prefix = 'https://autogluon.s3.amazonaws.com/datasets/Inc/'
    label = 'class'

    path_train = path_prefix + 'train.csv'
    path_test = path_prefix + 'test.csv'

    train_data = TabularDataset(path_train)
    train_data = train_data.sample(25000, random_state=0)
    test_data = TabularDataset(path_test)

    predictor: TabularPredictor = TabularPredictor(
        label=label,
        verbosity=3,
    ).fit(
        train_data=train_data,
        hyperparameters={'GBM': {}},
        # fit_weighted_ensemble=False,
        num_bag_folds=2,
        num_bag_sets=1,
    )
    predictor.leaderboard(test_data)

    from autogluon.core.metrics import accuracy, balanced_accuracy, f1, mcc, precision, recall
    y_pred = predictor.predict(data=test_data)
    result = predictor.evaluate_predictions(y_true=test_data[label], y_pred=y_pred)
    decision_threshold_dict = dict()
    for eval_metric in [accuracy, balanced_accuracy, f1, mcc, precision, recall]:
        decision_threshold = predictor.calibrate_decision_threshold(metric=eval_metric)
        decision_threshold_dict[eval_metric.name] = decision_threshold

        y_pred_w_decision_threshold = predictor.predict(data=test_data, decision_threshold=decision_threshold)
        result_calibrated = predictor.evaluate_predictions(y_true=test_data[label], y_pred=y_pred_w_decision_threshold)

        metric_score = result[eval_metric.name]
        metric_score_calibrated = result_calibrated[eval_metric.name]

        decision_threshold_dict[eval_metric.name] = dict(
            decision_threshold=decision_threshold,
            metric_score=metric_score,
            metric_score_calibrated=metric_score_calibrated,
        )

    for name, result_dict in decision_threshold_dict.items():
        decision_threshold = result_dict['decision_threshold']
        metric_score = result_dict['metric_score']
        metric_score_calibrated = result_dict['metric_score_calibrated']
        print(f'decision_threshold={decision_threshold:.3f}\t| metric="{name}"'
              f'\n\ttest_score uncalibrated: {metric_score:.4f}'
              f'\n\ttest_score   calibrated: {metric_score_calibrated:.4f}'
              f'\n\ttest_score        delta: {metric_score_calibrated-metric_score:.4f}')

```
</details>

<details>
  <summary>Example Output</summary>

```
decision_threshold=0.500	| metric="accuracy"
	test_score uncalibrated: 0.8703
	test_score   calibrated: 0.8703
	test_score        delta: 0.0000
decision_threshold=0.220	| metric="balanced_accuracy"
	test_score uncalibrated: 0.7925
	test_score   calibrated: 0.8476
	test_score        delta: 0.0551
decision_threshold=0.380	| metric="f1"
	test_score uncalibrated: 0.7022
	test_score   calibrated: 0.7245
	test_score        delta: 0.0222
decision_threshold=0.380	| metric="mcc"
	test_score uncalibrated: 0.6243
	test_score   calibrated: 0.6347
	test_score        delta: 0.0105
decision_threshold=0.990	| metric="precision"
	test_score uncalibrated: 0.7713
	test_score   calibrated: 1.0000
	test_score        delta: 0.2287
decision_threshold=0.000	| metric="recall"
	test_score uncalibrated: 0.6445
	test_score   calibrated: 1.0000
	test_score        delta: 0.3555
```
</details>

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

gradientsky commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -2507,6 +2526,112 @@ def fit_weighted_ensemble(self, base_models: list = None, name_suffix='Best', ex
+
+        if verbose:
+            logger.log(20, f'\nOptimizing Metric: {metric.name}')
+        for decision_threshold in decision_thresholds:
+            extra_log = ''
+            y_pred_val = get_pred_from_proba(
+                y_pred_proba=y_pred_proba_val.values,
+                problem_type=self.problem_type,
+                decision_threshold=decision_threshold,
+            )
+            # TODO: Avoid calling like this, re-use logic that works with weights + extra args
+            score_val = metric(y_true=y_val, y_pred=y_pred_val)
+
+            if best_score_val is None or score_val > best_score_val:
+                best_threshold = decision_threshold
+                best_score_val = score_val
+                extra_log = '\t| NEW BEST'
+            elif best_score_val == score_val:
+                # If the new threshold is closer to 0.5 than the previous threshold, prioritize it.
+                if abs(decision_threshold - 0.5) < abs(best_threshold - 0.5):
+                    best_threshold = decision_threshold
+                    best_score_val = score_val
+                    extra_log = '\t| NEW BEST (Tie, using threshold that is closer to 0.5)'
+
+            if verbose:
+                logger.log(15, f'\tthreshold: {decision_threshold:.3f}\t| val: {score_val:.4f}{extra_log}')
```
why don't we use bayesopt here. Something like this:
```python
from bayes_opt import BayesianOptimization

def score(self, threshold):
    # Implement your scoring function here
    # Make sure to pay attention if this greater is better here
    pass

optimizer = BayesianOptimization(
    f=score,
    pbounds={'threshold': (0, 1)},
    verbose=2, # verbose = 1 prints only when a maximum is observed, verbose = 0 is silent
    random_state=1,
)

optimizer.maximize(init_points=0, n_iter=decision_thresholds)

# Result
print("Best threshold: ", optimizer.max['params']['threshold'])
print("Best score: ", optimizer.max['target'])
```

Innixma replied:
We can leave this optimization for a future release, as the logic is quite fast already.

gradientsky commented on tabular/src/autogluon/tabular/learner/abstract_learner.py:
```diff
@@ -147,11 +147,23 @@ def predict_proba(self, X: DataFrame, model=None, as_pandas=True, as_multiclass=
+                inverse_transform=True,
+                transform_features=True,
+                *,
+                decision_threshold: float = None,
```
nit: `decision_threshold: Optional[float] = None`

Also, add types for other args above

Innixma replied:
We can add type hints in learner as part of a separate PR

gradientsky commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -1380,7 +1380,13 @@ def fit_pseudolabel(self, pseudo_data: pd.DataFrame, max_iter: int = 5, return_p
                                             **fit_extra_kwargs)
 
-    def predict(self, data, model=None, as_pandas=True, transform_features=True):
+    def predict(self,
+                data,
+                model=None,
+                as_pandas=True,
+                transform_features=True,
+                *,
+                decision_threshold: float = None):
```
nit: ditto (types)

Innixma replied:
We can add type hints in this method in a separate PR (Although note that docstring has the types listed)

gradientsky replied:
Sounds good;
I would note that types in docstrings are less helpful if we want to add automated types checks via mypy (i.e. like it's done in eda package).

Innixma replied:
added type hints

Innixma commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -2507,6 +2526,112 @@ def fit_weighted_ensemble(self, base_models: list = None, name_suffix='Best', ex
+            return 0.5
+
+        if model == 'best':
+            # FIXME: Incorrect if 'val' and model is refit_full
```
addressed this, refit_full works properly now, and only crashes if parent is deleted, otherwise the parent is used to calibrate.

Innixma commented:
I've added a major update to the PR.

Added logic:
- Support for leaderboard
- Support for evaluate / evaluate_predictions
- Support for predict_multi
- Support for .fit(…, calibrate_decision_threshold=True)
- Support for updating the default decision_threshold via predictor.set_decision_threshold 
- Support for refit_full models
- Support for `get_oof_pred`
- Tests for all of the above
- Various logging / docstrings

Innixma commented on core/src/autogluon/core/trainer/abstract_trainer.py:
```diff
@@ -3046,3 +3061,62 @@ def calibrate_model(self, model_name: str = None, lr: float = 0.01, max_iter: in
                 model.save()
+
+    def calibrate_decision_threshold(self,
+                                     X: pd.DataFrame = None,
+                                     y: np.array = None,
```
This is an interesting topic, because I almost feel that Optional is redundant when the default is None, as the Optional nature is self-evident by the default value.

For example, `mypi` and IDEs automatically convert it to the Optional type hint when default is None, meaning from a development and usage perspective on IDEs, it already hints correctly (https://stackoverflow.com/a/66922247).

The question is if we want the extra verbosity in our code. I am open to it, but I feel like it would be distracting more than helpful given that IDEs already auto-convert it, and we use None defaults everywhere, so it would be a ton of changes.

gradientsky replied:
mypy explicitly flags if you assign `None` to non-optional

gradientsky replied:
https://peps.python.org/pep-0484/#union-types
![image](https://github.com/autogluon/autogluon/assets/10080307/ef7181ca-6b54-4d6d-8840-f19fcd7f3909)

gradientsky replied:
Looks like [PEP604](https://peps.python.org/pep-0604/) confirms it's ok to replace `Optional[X]` with `X | None`.

Innixma replied:
Will use `X | None` with `from __future__ import annotations` following [PEP 604](https://peps.python.org/pep-0604/) instead of Optional as the pipe logic is more readable.

Innixma commented on core/src/autogluon/core/trainer/abstract_trainer.py:
```diff
@@ -3046,3 +3061,62 @@ def calibrate_model(self, model_name: str = None, lr: float = 0.01, max_iter: in
+            model = model_parent
+            if self.has_val:
+                # Use validation data
+                X = self.load_X_val()
+                if self.weight_evaluation:
+                    X, weights = extract_column(X=X, col_name=self.sample_weight)
+                y = self.load_y_val()
+                y_pred_proba = self.predict_proba(X=X, model=model)
+            else:
+                # Use out-of-fold data
+                if self.weight_evaluation:
+                    X = self.load_X()
```
Yes it is wiped when optimize for deployment, so it would crash. With that said, we don't have helpful logging in this situation for a lot of places, so doing it properly would be a non-trivial effort. I added a todo for this. I think most users won't run into this problem though, and specifying `calibrate_decision_threshold=True` in the fit call will ensure that the calibration occurs before the data is deleted.

Innixma commented on core/src/autogluon/core/utils/utils.py:
```diff
@@ -253,28 +253,44 @@ def augment_rare_classes(X, label, threshold):
+    elif problem_type == BINARY and decision_threshold is not None:
+        cols = y_pred_proba.columns
+        negative_class = cols[0]
+        positive_class = cols[1]
```
updated

Innixma commented on core/src/autogluon/core/trainer/abstract_trainer.py:
```diff
@@ -3046,3 +3061,62 @@ def calibrate_model(self, model_name: str = None, lr: float = 0.01, max_iter: in
+    def calibrate_decision_threshold(self,
+                                     X: pd.DataFrame = None,
+                                     y: np.array = None,
+                                     metric: Union[str, Scorer] = None,
```
added

Innixma commented on core/src/autogluon/core/utils/utils.py:
```diff
@@ -253,28 +253,44 @@ def augment_rare_classes(X, label, threshold):
-    """From input DataFrame of pred_proba, return Series of pred"""
+def get_pred_from_proba_df(y_pred_proba: pd.DataFrame,
+                           problem_type: str = BINARY,
+                           decision_threshold: float = None) -> pd.Series:
```
added

Innixma commented:
Benchmark results show no changes from this PR, as expected.

rafijacSense commented:
Would be great if this was included in MultiModalPredictor
